qp_klp/klp.py

```diff
@@ -121,6 +123,43 @@ def sequence_processing_pipeline(qclient, job_id, parameters, out_dir):
         for sample in sheet:
             sample['Lane'] = f'{lane_number}'
 
+        sheet_df = sample_sheet_to_dataframe(sheet)
+        errors = []
+        for project, _df in sheet_df.groupby('sample_project'):
+            project_name = remove_qiita_id(project)
+            qiita_id = project.replace(f'{project_name}_', '')
+            qurl = f'/api/v1/study/{qiita_id}/samples'
+            sheet_samples = {
+                s for s in _df['sample_name'] if not s.startswith('BLANK')}
+            qsamples = {
+                s.replace(f'{qiita_id}.', '') for s in qclient.get(qurl)}
+            sample_name_diff = sheet_samples - qsamples
+            if sample_name_diff:
+                # before we report as an error, check tube_id
+                error_tube_id = 'No tube_id column in Qiita.'
+                if 'tube_id' in qclient.get(f'{qurl}/info')['categories']:
+                    tids = qclient.get(f'{qurl}/categories=tube_id')['samples']
+                    tids = {tid[0] for _, tid in tids.items()}
+                    tube_id_diff = sheet_samples - tids
+                    if not tube_id_diff:
+                        continue
+                    len_tube_id_overlap = len(tube_id_diff)
+                    tids_example = ', '.join(choices(list(tids), k=5))
+                    error_tube_id = (
+                        f'tube_id in Qiita but {len_tube_id_overlap} missing '
+                        f'samples. Some samples from tube_id: {tids_example}.')
+
+                len_overlap = len(sample_name_diff)
+                samples_example = ', '.join(choices(list(qsamples), k=5))
```

Comment:
k=5 and :4 seem like magic numbers to me. It would be nice to add a comment on what they represent.

Reply:
ty!